[quant] Implement APoT_tensor class

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79845
* __->__ #79940

### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and a method to return the int representation of an APoT tensor.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

asl3 committed:
[quant] Implement APoT_tensor class

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -2,13 +2,13 @@
     def __init__(self, quantizer):
-        raise NotImplementedError
+        self.quantizer = quantizer
+        self.data = quantizer.data
```
feels like the data should be stored in the tensor rather than in the quantizer, i would probably refactor the quantizer to take the data as an argument for this PR

asl3 replied:
fixed!

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -2,11 +2,11 @@
 
 # class to store APoT quantized tensor
-class TensorAPoT(torch.Tensor):
+class TensorAPoT():
```
there was a comment in the previous PRs suggesting to store the quantized tensor in TensorAPot, but I don't see this being done in this PR?

dzdang replied:
TensorAPoT should store the quantized tensor i.e., the output of your `quantize_apot` method

asl3 replied:
@dzdang yep but `quantize_apot` needs a tensor2quantize and quantizer doesn't store data -- i guess then tensor2quantize should get passed in to my `init` function for apot_tensor?

dzdang replied:
not sure what the issue is? quantize_apot returns the quantized tensor which you can store in apot_tensor

asl3 replied:
should the parameter that `quantize_apot` needs be passed to it by the `init` function in apot_tensor?

dzdang replied:
yes that works or we can do this in a 2 step process:
1) construct apot tensor with quantizer and leave data field uninitialized
2) create a quantize method in apot_tensor that takes in fp tensor that calls quantizer->quantize_apot

asl3 replied:
ok sounds great, will do

dzdang commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -2,11 +2,11 @@
-    def int_repr(self):
-        raise NotImplementedError
+    def int_repr(self, tensor: torch.Tensor):
+        return self.quantizer.quantize_APoT(tensor)
```
see above comment and https://pytorch.org/docs/stable/generated/torch.Tensor.int_repr.html. 
`int_repr` is a getter function. it shouldn't be quantizing anything

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and has a method to convert APoT tensor to int representation.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement APoT_tensor class"


### Summary:
This PR implements functionality to store APoT tensors. The APoT tensor class contains a quantizer and a method to return the int representation of an APoT tensor.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -2,11 +2,13 @@
 
 # class to store APoT quantized tensor
-class TensorAPoT(torch.Tensor):
+class TensorAPoT():
```
why is TensorAPoT no longer inheriting from torch.Tensor?

asl3 replied:
When TensorAPoT inherits from torch.Tensor I get an issue with my TensorAPoT `init` function since I am not passing in the expected args:
`TypeError: new() received an invalid combination of arguments - got (Tensor, int, int, bool), but expected one of:
 * (*, torch.device device)
 * (torch.Storage storage)
 * (Tensor other)
 * (tuple of ints size, *, torch.device device)
 * (object data, *, torch.device device)`

I didn't think any of these made sense for my TensorAPoT object since I am passing in a quantizer object. Closest was the `(Tensor other)` option, but in that case I would need to create the quantizer in my `init` function and would need to pass in the args needed by quantizer `init` , which would also violate the expected args.

Is it okay if TensorAPoT doesn't inherit from torch.Tensor?

jerryzh168 replied:
I feel this is fine for now, this might involve a bit more design about how do we allow people to extend quantized tensor in python, that's something we can explore in H2
